### DISCUSSION TITLE
trigger routing manually

### DISCUSSION THREAD

**Discussion by bluebrown** [fastapi/fastapi#4352] [Questions]
Hi, I want to implement a batch handler that takes a request with multipart/mixed content type. Each part is its own sub request. Hence, I am looking for a way to trigger a routing for each subrequest.

The batch request would have the following format.

```http
POST /batch  HTTP/1.1
Host: localhost:8080
Content-Type: multipart/mixed; boundary=boundary

--boundary
Content-Type: application/http
Content-ID: <item1>

POST /items HTTP/1.1
Content-Type: application/json
Content-length: 58
Accept: application/json

{ "name": "batch1", "description": "batch1 description" }

--boundary
Content-Type: application/http
Content-ID: <item2>

PUT /other HTTP/1.1
Content-Type: application/json
Content-length: 58
Accept: application/json

{ "name": "batch2", "description": "batch1 description" }

--boundary--
```

So, I am looking for a way to take each subrequest and dispatch it to fastapi so that it routes to the corresponding endpoint.

Something like the below pseudocode.

```python
def batch() -> None:
    for subrequest in request:
        response = FastAPI.handle_request(subrequest)
```

I have implemented this as proof of concept in go https://gist.github.com/bluebrown/f0401fd7c1941985101a35085c041502. But now I am looking to get this working in fastapi.

**Comment by sasano8** [comment]
I'm interested in that idea.

I hope I can implement a batch request like JSONRPC.

The following code is not enough to meet your requirements, but you can request it from any endpoint.

After that, you can implement your idea by finding a way to combine multiple results.

``` python
from fastapi import FastAPI
from fastapi.routing import APIRoute, APIRouter
from fastapi.testclient import TestClient
from fastapi import HTTPException
from fastapi.responses import PlainTextResponse


class MyRoute(APIRoute):
    async def handle(self, scope, receive, send) -> None:
        if self.methods and scope["method"] not in self.methods:
            if "app" in scope:
                raise HTTPException(status_code=405)
            else:
                response = PlainTextResponse("Method Not Allowed", status_code=405)
            await response(scope, receive, send)
        else:
            if scope["path"] == "/":
                scope["path"] += "hello"
                await scope["router"].__call__(scope, receive, send)
            else:
                await self.app(scope, receive, send)


app = FastAPI()
router = APIRouter(route_class=MyRoute)

@router.get("/")
def dummy_entrypoint():
    ...

@router.get("/hello")
def hello():
    return "hello"

app.include_router(router)

client = TestClient(app)

response = client.get("/")
assert {"status_code": response.status_code, "result": response.json(),} == {
    "status_code": 200,
    "result": "hello",
}
```

**Reply by bluebrown** [reply]
I just thought, what if we would hook into the asgi server thats serving the fastapi app, to split the request and dispatch it to fastapi. for example when using uvivorn.

**Reply by sasano8** [reply]
There is only one response per http connection.

By adding `more_body` at the time of transmission, the response can be divided and returned.
Otherwise, the connection will be closed as if the process was completed.

- [Send event]https://asgi.readthedocs.io/en/latest/specs/www.html
